python/ray/data/dataset.py

```diff
@@ -2172,14 +2181,14 @@ def zip(self, other: "Dataset[U]") -> "Dataset[(T, U)]":
         if logical_plan is not None and other_logical_plan is not None:
             op = Zip(logical_plan.dag, other_logical_plan.dag)
             logical_plan = LogicalPlan(op)
-        return Dataset(plan, self._epoch, self._lazy, logical_plan)
+        return Datastream(plan, self._epoch, self._lazy, logical_plan)
 
     @ConsumptionAPI
-    def limit(self, limit: int) -> "Dataset[T]":
-        """Truncate the dataset to the first ``limit`` records.
+    def limit(self, limit: int) -> "MaterializedDatastream[T]":
```

Comment:
Can we keep the return value as Datastream instead of MaterializedDatastream? We will soon implement the union as a streaming physical operator (https://github.com/ray-project/ray/issues/32761) during Ray 2.5.